### PR TITLE
Run ClassDefChecker after the Optimizer

### PR DESCRIPTION
**Pull request by gzm0**

### CODE REVIEW

File: linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala [hunk 1]
```diff
@@ -644,18 +649,19 @@ private final class ClassDefChecker(classDef: ClassDef, reporter: ErrorReporter)
...
-        if (!array.tpe.isInstanceOf[ArrayType])
-          reportError(i"Array type expected but ${array.tpe} found")
+        checkArrayReceiverType(array.tpe)
         checkTree(array, env)
         checkTree(index, env)
 
-      case _:RecordSelect | _:RecordValue =>
-        reportError("invalid tree")
+      case RecordSelect(record, _) if allowTransients =>
+        checkTree(record, env)
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:661]: Consider something like the following instead:
```suggestion
      case RecordSelect(record, _) =>
        if (!allowTransients)
          reportError("invalid tree")
        checkTree(record, env)
```
Then you can remove the default `case _: ... =>` at the end.

That duplicates the error message a bit, but it improves the exhaustivity check for the whole `match`, as well as keep checking subtrees even if we find a bad one.

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/Refiner.scala [hunk 2]
```diff
@@ -133,22 +136,31 @@ private object Refiner {
...
+    private def update(classDef: ClassDef, logger: Logger): Unit = synchronized {
       if (!cacheUsed) {
         cacheUsed = true
 
+        if (checkIR) {
+          val errorCount = ClassDefChecker.check(classDef,
+              allowReflectiveProxies = true, allowTransients = true, logger)
+          if (errorCount != 0) {
+            throw new AssertionError(s"There were $errorCount ClassDef " +
+                "checking errors after optimizing. Please report this as a bug.")
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/Refiner.scala:160]: To keep sentences on a single line (better grepping):
```suggestion
            throw new AssertionError(
                s"There were $errorCount ClassDef checking errors after optimizing. " +
                "Please report this as a bug.")
```

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala [hunk 3]
```diff
@@ -5247,7 +5241,18 @@ private[optimizer] object OptimizerCore {
...
+         * it also creates inconsistent trees:
+         * - This() not typed as the enclosing class.
+         * - VarRef not typed as the corresponding VarDef / ParamDef.
+         *
+         * Type based optimizations happen (mainly) in the optimizer so
+         * consistent downstream types are more important than narrower types;
+         * notably because it allows us to run the ClassDefChecker after the
+         * optimizer.
+         */
+        localDef.newReplacement
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:5255]: Does this change result in any different in the produced `.js` for the test suite?

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala [hunk 4]
```diff
@@ -731,12 +731,12 @@ private[optimizer] abstract class OptimizerCore(
...
         val localDef = LocalDef(tcaptureValue.tpe, mutable, replacement)
         val localIdent = LocalIdent(newName)(ident.pos)
-        val newParamDef = ParamDef(localIdent, newOriginalName, ptpe, mutable)(paramDef.pos)
+        val newParamDef = ParamDef(localIdent, newOriginalName, tcaptureValue.tpe.base, mutable)(paramDef.pos)
 
         /* Note that the binding will never create a fresh name for a
          * ReplaceWithVarRef. So this will not put our name alignment at risk.
          */
-        val valueBinding = Binding.temp(paramName, ptpe, mutable, tcaptureValue)
+        val valueBinding = Binding.temp(paramName, tcaptureValue.tpe.base, mutable, tcaptureValue)
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:739]: Given that `mutable == false` (per assertion), this is now equivalent to
```suggestion
        val valueBinding = Binding.temp(paramName, tcaptureValue)
```

**Review by sjrd — CHANGES_REQUESTED** [review]
Looks nice. I expected more changes required in the optimizer, so it's good to see we were not too far away (anymore).

File: linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala (hunk 1 above)
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/checker/ClassDefChecker.scala:661]: Good point. I ended up making a `checkAllowTransients()` which gives us the best of both worlds :)

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala (hunk 3 above)
Comment (gzm0) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala:5255]: Nope. They are byte by byte identical.

**Review by sjrd — APPROVED** [review]

**Pull request merged**